[all] Adds forward declaration and opaque api for the base types (the one used in Node.h)

damienmarchal opened the pull request:
This is the second step of removal of Base* inclusion in Node.h 
This is made possible by all the PR that layed out the fwd stuff and Link refactoring. 



______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

damienmarchal committed:
[Sofa.Core] Add forward declaration & core::opaque functions

The forward declaration are for commonly used type in our code base.

damienmarchal committed:
[SofaCore] Use opaque function in PathResolver.h and Link.h

damienmarchal committed:
[all] Add needed includes in .cpp, removes un-needed ones in .h

damienmarchal committed:
SofaCore] Move sofa::core::base::GetClass into sofa::core::objectmodel::base::GetClass

Because there should be a 1 to 1 mapping between namespace::ObjectName::functionName
and their opaque version namespace::objectname::functionName

fredroy commented on SofaKernel/modules/SofaCore/src/sofa/core/fwd.cpp:
```diff
@@ -1,5 +1,40 @@
+#include <sofa/core/visual/Shader.h>
+
+#include <sofa/core/collision/Pipeline.h>
+#include <sofa/core/collision/Pipeline.h>
```
You really want to include Pipeline.h 🧐

damienmarchal committed:
[SofaCore] FIX PR to follow code review.

Rename macro into:
- SOFA_DEFINE_OPAQUE_FUNCTIONS_BETWEEN_BASE_AND()
- SOFA_DECLARE_OPAQUE_FUNCTIONS_BETWEEN_BASE_AND()

Rename baseFrom/castBastTo into
- dynamicCastBaseFrom()
- dynamicCastBaseTo()

Remove duplicated Pipeline

Add docs for the macro's

fredroy commented:
Corrections from reviews done, ready to ~~rumble~~ merge
